g2p_programs/models/managers/entitlement_manager.py

```diff
@@ -60,6 +63,70 @@ def approve_entitlements(self, entitlements):
         """
         raise NotImplementedError()
 
+    def cancel_entitlements(self, cycle):
+        """Base Entitlement Manager :meth:`cancel_entitlements`
+        Cancel entitlements in a cycle
+        Override in entitlement manager
+
+        :param cycle: A recordset of cycle
+        :return:
+        """
+
+    def _cancel_entitlements_async(self, cycle, entitlements_count):
+        """Cancel Entitlements
+        Base Entitlement Manager :meth:`_cancel_entitlements_async`
+        Asynchronous cancellation of entitlements in a cycle using `job_queue`
+
+        :param cycle: A recordset of cycle
+        :param entitlements_count: Integer value of total entitlements to process
+        :return:
+        """
+        _logger.debug("Cancel entitlements asynchronously")
+        cycle.message_post(
+            body=_("Cancel %s entitlements started.", entitlements_count)
+        )
+        cycle.write(
+            {
+                "locked": True,
+                "locked_reason": _("Cancel entitlements for cycle."),
+            }
+        )
+
+        jobs = []
+        for i in range(0, entitlements_count, 2000):
+            jobs.append(self.delayable()._cancel_entitlements(cycle, i, 2000))
+        main_job = group(*jobs)
+        main_job.on_done(
+            self.delayable().mark_job_as_done(cycle, _("Entitlements Cancelled."))
+        )
+        main_job.delay()
+
+    def _cancel_entitlements(self, cycle, offset=0, limit=None):
+        """
+        Base Entitlement Manager :meth:`_cancel_entitlements`
+        Synchronous cancellation of entitlements in a cycle
+        Override in entitlement manager
+
+        :param cycle: A recordset of cycle
+        :param offset: An integer value to be used in :meth:`cycle.get_entitlements` for setting the query offset
+        :param limit: An integer value to be used in :meth:`cycle.get_entitlements` for setting the query limit
+        :return:
+        """
```

Comment:
Is `pass` missing ?

Reply:
Fixed

Reply:
pass is removed by pre-commit in function without codes. I changed it to `raise NotImplementedError()`